src/ng-select/items-list.ts

```diff
@@ -305,11 +304,29 @@ export class ItemsList {
     }
 
     private _groupBy(items: NgOption[], prop: string | Function): OptionGroups {
-        const isFn = isFunction(this._ngSelect.groupBy);
-        const groups = new Map<string, NgOption[]>();
+        const groups = new Map<string | NgOption, NgOption[]>();
+        if (items.length === 0) {
+            return groups;
+        }
+
+        // Check if items are already grouped by given key.
+        if (Array.isArray(items[0].value[<string>prop])) {
```

Comment:
I thought about that, but since it need to map to new Map<string | NgOption, NgOption[]>(); which declared already inside _groupBy let's keep it in one place for now.